lightning/src/routing/scoring.rs

```diff
@@ -3725,6 +3903,59 @@ mod tests {
 		assert_eq!(scorer.historical_estimated_payment_success_probability(42, &target, amount_msat, &params, false),
 			Some(0.0));
 	}
+
+	#[test]
+	fn combined_scorer() {
+		let logger = TestLogger::new();
+		let network_graph = network_graph(&logger);
+		let params = ProbabilisticScoringFeeParameters::default();
+		let mut scorer = ProbabilisticScorer::new(ProbabilisticScoringDecayParameters::default(), &network_graph, &logger);
+		scorer.payment_path_failed(&payment_path_for_amount(600), 42, Duration::ZERO);
+
+		let mut combined_scorer = CombinedScorer::new(scorer);
+
+		// Verify that the combined_scorer has the correct liquidity range after a failed 600 msat payment.
+		let liquidity_range = combined_scorer.scorer.estimated_channel_liquidity_range(42, &target_node_id());
+		assert_eq!(liquidity_range.unwrap(), (0, 600));
+
+		let source = source_node_id();
+		let usage = ChannelUsage {
+			amount_msat: 750,
+			inflight_htlc_msat: 0,
+			effective_capacity: EffectiveCapacity::Total { capacity_msat: 1_000, htlc_maximum_msat: 1_000 },
+		};
+		
+		{
+			let network_graph = network_graph.read_only();
+			let channel = network_graph.channel(42).unwrap();
+			let (info, _) = channel.as_directed_from(&source).unwrap();
+			let candidate = CandidateRouteHop::PublicHop(PublicHopCandidate {
+				info,
+				short_channel_id: 42,
+			});
+			
+			let penalty = combined_scorer.channel_penalty_msat(&candidate, usage, &params);
+			
+			let mut external_liquidity = ChannelLiquidity::new(Duration::ZERO);
+			let logger_rc = Rc::new(&logger);	// Why necessary and not above for the network graph?
+			external_liquidity.as_directed_mut(&source_node_id(), &target_node_id(), 1_000).
+				successful(1000, Duration::ZERO, format_args!("test channel"), logger_rc.as_ref());
+
+			let mut external_scores = ChannelLiquidities::new();
+		
+			external_scores.insert(42, external_liquidity);
+			combined_scorer.merge(external_scores, Duration::ZERO);
+
+			let penalty_after_merge = combined_scorer.channel_penalty_msat(&candidate, usage, &params);
+			
+			// Since the external source observed a successful payment, the penalty should be lower after the merge.
+			assert!(penalty_after_merge < penalty);
+		}
+
+		// Verify that after the merge with a successful payment, the liquidity range is increased.
+		let liquidity_range = combined_scorer.scorer.estimated_channel_liquidity_range(42, &target_node_id());
+		assert_eq!(liquidity_range.unwrap(), (0, 300));
```

Comment:
I don't think this number is too magical. Just sitting in between the previous max offset of 600 and the new offset of 0.